doc/architecture-decisions/0016-job-for-upgrading-istio-sidecars-on-workloads.md

```diff
@@ -0,0 +1,55 @@
+# 16. Job for Upgrading Istio Sidecars on Workloads
+
+Date: 2020-08-11
+
+## Status
+
+Accepted
+
+## Context
+
+Istio's service mesh capabilites are facilitated via sidecars injected into
+workload pods. These sidecars run an Istio-patched version of Envoy that is tied
+to the version of Istio that injects them.
+
+Typically when new versions of Istio are released, new versions of the sidecars
+are released as well. Istio has been good so far about supporting older versions
+of sidecars that were deployed before Istio was upgraded, but it is still
+[documented best practice](https://istio.io/latest/docs/setup/upgrade/) to roll
+all the pods after an Istio upgrade.
+
+As an additional constraint, the operators of cf-for-k8s clusters expect to be
+able to perform upgrades in one `kapp deploy`, with no post-install hooks or
+other bash scripts. This limits our options considerably. See this [Slack
+thread](https://cloudfoundry.slack.com/archives/CH9LF6V1P/p1592521879117400) on
+that constraint.
+
+## Decision
+
+We will use the kubernetes
+[Job](https://kubernetes.io/docs/concepts/workloads/controllers/job/) resource
+to run the kubectl command needed to roll workload pods, after waiting for the
+new Istio control plane to be up and healthy.
+
+To that end, we will add the necessary minimal `ServiceAccounts` and `Roles`
+needed to list resources in the `istio-system` namespace, and restart resources
+in the configured workload namespace. We will also build and maintain a
+container image that contains the Job's logic.
+
+All istio components will be tagged with their Istio version so that the job can
+positively determine that the correct version of control plane components are
+alive and healthy. We will also name the job according to it's Istio version, so
+that we can take advantage of `Jobs` inherent immutability in cases where a
+cf-for-k8s upgrade does not contain a new Istio version (pushing the same job
+again will not cause it to rerun, preventing workloads from rolling
+unnecessarily). Subsequent jobs will clean up previous ones.
+
```

Comment:
Maybe we could add a bit about how the Job gets triggered, and how it waits on the labeled istio resources to reach a new version before then restarting the workloads. One thing that I keep having to remind myself about is that we wait for 2 configmaps before restarting the workloads, and would be nice to capture that here.

Reply:
We don't do that any more! See https://github.com/cloudfoundry/cf-k8s-networking/commit/cbade83a3c344918c1b580e1ebedb0a6d986c345

I wonder how nitty gritty we want to be in these ADRs. I worry about too closely coupling the ADR to details of the implementation that may change over time and the ADR becoming stale. This is a good example of that, and we may later find a better way of achieving our goal of waiting util control plane components are ready. I wouldn't want to have to come back and change the ADR in that case, so I tried to phrase it in more of an outcome centered way than an implementation way.

Reply:
Ah oops! And that totally makes sense. Resolving